tutorial/lesson_17_predicated_rdom.cpp

```diff
@@ -0,0 +1,233 @@
+            for (int r_x = 0; r_x < 8; r_x++) {
+                // Update is only performed if the predicate evaluates to true.
+                if ((r_x + r_y > 5) && (3*r_y - 2*r_x < 15) && (4*r_x - r_y < 20)) {
+                    c_result[r_y][r_x] *= 2;
+                }
+            }
+        }
+
+        // Check the results match:
+        for (int y = 0; y < 10; y++) {
+            for (int x = 0; x < 10; x++) {
+                if (halide_result(x, y) != c_result[y][x]) {
+                    printf("halide_result(%d, %d) = %d instead of %d\n",
+                           x, y, halide_result(x, y), c_result[y][x]);
+                    return -1;
+                }
+            }
+        }
+    }
+
+    {
+        // The predicate is not limited to RDom's variables only (e.g. r.x, r.y, etc.).
+        // It could refer to free variables in the update definition or calls to
+        // other Funcs, or make recursive calls to the same Func. Here is some
+        // example:
+        Func f("call_f"), g("call_g");
+        Var x("x"), y("y");
+        f(x, y) = 2 * x + y;
+        g(x, y) = x + y;
+
+        // This RDom's predicates depend on the initial value of 'f'.
+        RDom r1(0, 5, 0, 5, "r1");
+        r1.where(f(r1.x, r1.y) >= 10);
+        r1.where(f(r1.x, r1.y) != 15);
+        f(r1.x, r1.y) += 1;
+        f.compute_root();
+
+        // While this one involves calls to another Func.
+        RDom r2(1, 3, 1, 3, "r2");
+        r2.where(f(r2.x, r2.y) < 15);
+        g(r2.x, r2.y) += f(r2.x, r2.y)/2;
+
+        Buffer<int> halide_result_f = f.realize(5, 5);
+        Buffer<int> halide_result_g = g.realize(5, 5);
+
+        // See figures/lesson_17_rdom_calls_in_predicate.gif for a visualization
+        // of what this did.
+
+        // The equivalent C for 'f' is:
+        int c_result_f[5][5];
+        for (int y = 0; y < 5; y++) {
+            for (int x = 0; x < 5; x++) {
+                c_result_f[y][x] = 2 * x + y;
+            }
+        }
+        for (int r1_y = 0; r1_y < 5; r1_y++) {
+            for (int r1_x = 0; r1_x < 5; r1_x++) {
+                // Update is only performed if the predicate evaluates to true.
+                if ((c_result_f[r1_y][r1_x] >= 10) && (c_result_f[r1_y][r1_x] != 15)) {
+                    c_result_f[r1_y][r1_x] += 1;
+                }
+            }
+        }
+
+        // And, the equivalent C for 'g' is:
+        int c_result_g[5][5];
+        for (int y = 0; y < 5; y++) {
+            for (int x = 0; x < 5; x++) {
+                c_result_g[y][x] = x + y;
+            }
+        }
+        for (int r2_y = 1; r2_y < 4; r2_y++) {
+            for (int r1_x = 1; r1_x < 4; r1_x++) {
+                // Update is only performed if the predicate evaluates to true.
+                if (c_result_f[r2_y][r1_x] < 15) {
+                    c_result_g[r2_y][r1_x] += c_result_f[r2_y][r1_x]/2;
+                }
+            }
+        }
+
+        // Check the results match:
```

Comment:
This is supposed to be the equivalent C loop structure, but for_each_element leaves the loop structure implicit, so it wouldn't really convey what this is trying to convey.